pkg/bus/bus.go

```diff
@@ -35,26 +36,33 @@ func (eb *EventBus) Publish(e data.Event) {
 	eb.rw.RUnlock() //defer is actually very slow
 }
 
+//RecieveFunc callback type for receiving metrics
+// Arguments are name, timestamp, metric type, interval, value, labels
+type RecieveFunc func(string, time.Time, data.MetricType, time.Duration, float64, []string, []string)
+
+//PublishFunc ...
+type PublishFunc func(string, time.Time, data.MetricType, time.Duration, float64, []string, []string)
+
 //MetricBus bus for data.Metric type
 type MetricBus struct {
-	subscribers []chan []data.Metric
-	rw          sync.RWMutex
+	sync.RWMutex
+	subscribers []RecieveFunc
 }
 
 //Subscribe subscribe to bus
-func (mb *MetricBus) Subscribe(c chan []data.Metric) {
-	mb.rw.Lock()
-	defer mb.rw.Unlock()
-	mb.subscribers = append(mb.subscribers, c)
+func (mb *MetricBus) Subscribe(rf RecieveFunc) {
+	mb.Lock()
+	defer mb.Unlock()
+	mb.subscribers = append(mb.subscribers, rf)
 }
 
 //Publish publish to bus
-func (mb *MetricBus) Publish(m []data.Metric) {
-	mb.rw.RLock()
-	for _, c := range mb.subscribers {
-		go func(c chan []data.Metric, m []data.Metric) {
-			c <- m
-		}(c, m)
+func (mb *MetricBus) Publish(name string, time time.Time, typ data.MetricType, interval time.Duration, value float64, labelKeys []string, labelVals []string) {
```

Comment:
Agreed, it is a lot uglier than the original which is unfortunate. The original design used the data.Metric type, which had all of the same fields within it as the function signature for the callback. I tried to standardize it so that metrics from pretty much anywhere would fit in. So: name field, time stamp, interval, value, and labels. Metadata fields may be worth adding. Ceilometer metrics is next so if I find that necessary I will add it.

I didn't try any third party queues. Anything that involves using the network stack of the OS I didn't even consider testing because I don't think there is any way that would be faster than keeping everything at a lower level within the golang runtime and thus did not take the time to implement and test. There would have to be additional processing done to re-marshal the incoming message into a byte string representation of the data just to send it to another sg-core component that would then have to unmarshal it again. That would be significantly more overhead, and based on my cpu profiling, the unmarshalling already takes up a fair amount of time. Granted, queueing would help but I still don't think it could be faster. Maybe that is something we could test on a smaller scale and consider transitioning to in the future if it is worth it. 

What I could try is Aaron's suggestion of using non-blocking channels in the internal bus and just drop unprocessed messages. This could potentially be even faster and, since we really only care about the metric when Prometheus scrapes once every second, there is no real reason to process every single incoming message anyway except those that get there in time to be scrapped. This is my opinion, however, and maybe this is not the functionality we want. I think, in this situation, we would want to be careful that all metrics are dropped evenly to avoid staleness problems of one metric. Maybe I'll try testing that approach today